File: protocol/python/job.py
from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True, slots=True)
class JobDescriptionExtension:
    job_title: str
    source_url: str

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "JobDescriptionExtension":
        return cls(
            job_title=str(data["job_title"]),
            source_url=str(data["source_url"]),
        )

    def to_dict(self) -> dict[str, Any]:
        return {
            "job_title": self.job_title,
            "source_url": self.source_url,
        }


@dataclass(frozen=True, slots=True)
class JobDescriptionList:
    jobs: list[JobDescriptionExtension]

    @classmethod
    def from_list(cls, data: list[dict[str, Any]]) -> "JobDescriptionList":
        return cls(jobs=[JobDescriptionExtension.from_dict(item) for item in data])

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "JobDescriptionList":
        if "jobs" not in data:
            raise ValueError("Expected a dict with 'jobs' key")
        return cls.from_list(data["jobs"])

    @classmethod
    def from_json(cls, data: str) -> "JobDescriptionList":
        parsed = json.loads(data)
        if isinstance(parsed, dict) and "jobs" in parsed:
            return cls.from_dict(parsed)
        if isinstance(parsed, list):
            return cls.from_list(parsed)
        raise ValueError("Expected a JSON object with 'jobs' key or a JSON array")

    def to_list(self) -> list[dict[str, Any]]:
        return [job.to_dict() for job in self.jobs]

    def to_dict(self) -> dict[str, Any]:
        return {"jobs": self.to_list()}

    def to_json(self) -> str:
        return json.dumps(self.to_dict(), ensure_ascii=False)

File: protocol/js/job.ts
interface JobDescriptionExtensionData {
  job_title: string;
  source_url: string;
}

interface JobDescriptionListData {
  jobs: JobDescriptionExtensionData[];
}

class JobDescriptionExtension {
  readonly job_title: string;
  readonly source_url: string;

  constructor({ job_title, source_url }: JobDescriptionExtensionData) {
    this.job_title = String(job_title);
    this.source_url = String(source_url);
  }

  static fromDict(data: JobDescriptionExtensionData): JobDescriptionExtension {
    return new JobDescriptionExtension({
      job_title: data.job_title,
      source_url: data.source_url,
    });
  }

  toDict(): JobDescriptionExtensionData {
    return {
      job_title: this.job_title,
      source_url: this.source_url,
    };
  }
}

class JobDescriptionList {
  readonly jobs: JobDescriptionExtension[];

  constructor(jobs: JobDescriptionExtension[]) {
    this.jobs = jobs;
  }

  static fromList(data: JobDescriptionExtensionData[]): JobDescriptionList {
    if (!Array.isArray(data)) {
      throw new Error("Expected an array of job descriptions");
    }

    return new JobDescriptionList(data.map((item) => JobDescriptionExtension.fromDict(item)));
  }

  static fromDict(data: JobDescriptionListData): JobDescriptionList {
    if (!("jobs" in data)) {
      throw new Error("Expected an object with 'jobs' key");
    }
    return JobDescriptionList.fromList(data.jobs);
  }

  static fromJson(data: string): JobDescriptionList {
    const parsed = JSON.parse(data);
    if (typeof parsed === "object" && parsed !== null && "jobs" in parsed) {
      return JobDescriptionList.fromDict(parsed as JobDescriptionListData);
    }
    if (Array.isArray(parsed)) {
      return JobDescriptionList.fromList(parsed);
    }
    throw new Error("Expected a JSON object with 'jobs' key or a JSON array");
  }

  toList(): JobDescriptionExtensionData[] {
    return this.jobs.map((job) => job.toDict());
  }

  toDict(): JobDescriptionListData {
    return { jobs: this.toList() };
  }

  toJson(): string {
    return JSON.stringify(this.toDict());
  }
}

export { JobDescriptionExtensionData, JobDescriptionListData, JobDescriptionExtension, JobDescriptionList };
